File: SocketUnix/Esame/largest_expenses.java
import java.io.*;
import java.net.*;

public class largest_expenses
{
    public static void main(String args[])
    {
        if (args.length != 2) {
            System.err.println("Errore! La sintassi corretta è: java largest_expenses hostname porta");
            System.exit(1);
        }

        try {
            Socket theSocket = new Socket(args[0], Integer.parseInt(args[1]));

            BufferedReader userIn = new BufferedReader(new InputStreamReader(System.in));
            BufferedReader networkIn = new BufferedReader(new InputStreamReader(theSocket.getInputStream(),"UTF-8"));
            BufferedWriter networkOut = new BufferedWriter(new OutputStreamWriter(theSocket.getOutputStream(),"UTF-8"));

            for (;;) {
                System.out.println("Inserisci l'anno di interesse, 'fine' per uscire: ");
                String anno = userIn.readLine();
                if (anno.equals("fine")){
                    System.out.println("Hai scelto di terminare il programma.");
                    break;
                }
                try{
                    Integer.parseInt(anno);
                }catch(NumberFormatException e){
                    System.err.println("Inserire un anno valido");
                    break;
                }
                networkOut.write(anno);
                networkOut.newLine();
                networkOut.flush();
                //CONTROLLO ACK
                String buff = networkIn.readLine();
                if(!buff.equals("ack")){
                    System.err.println("Errore lettura Ack dal server");
                    System.exit(2);
                }
                //Numero di spese
                System.out.println("Inserisci il numero di spese di interesse, 'fine' per uscire: ");
                String N = userIn.readLine();
                if (N.equals("fine")){
                    System.out.println("Hai scelto di terminare il programma.");
                    break;
                }
                try{
                    Integer.parseInt(N);
                }catch(NumberFormatException e){
                    System.err.println("Inserire un numero valido");
                    break;
                }
                networkOut.write(N);
                networkOut.newLine();
                networkOut.flush();

                //CONTROLLO ACK
                buff = networkIn.readLine();
                if(!buff.equals("ack")){
                    System.err.println("Errore lettura Ack dal server");
                    System.exit(2);
                }
                //Tipolgia di spese
                System.out.println("Inserisci la tipologia di spese, 'fine' per uscire: ");
                String tipologia = userIn.readLine();
                
                if (tipologia.equals("fine")){
                    System.out.println("Hai scelto di terminare il programma.");
                    break;
                }
                networkOut.write(tipologia);
                networkOut.newLine();
                networkOut.flush();

                for (;;) {
                    //Leggo input da Server riga per riga 
                    buff = networkIn.readLine();

                    /* Controllo errori */
                    if (buff == null) {
                        System.err.println("Errore! Il Server ha chiuso la connessione!");
                        System.exit(2);
                    }

                    /* Stampo riga letta da Server */
                    System.out.println(buff);

                    /* Passo a nuova richiesta una volta terminato input Server */
                    if (buff.equals("--END--")) {
                        break;
                    }
                }
            }

            /* chiudo la socket! */
            theSocket.close();
        }
        catch (IOException e) {
            System.err.println(e.getMessage());
            e.printStackTrace();
            System.exit(2);
        }
    }
}

File: SocketUnix/Esame/largest_expenses.c
#define _POSIX_C_SOURCE 200809L
#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "rxb.h"
#include "utils.h"

/* La massima dimensione di una richiesta è di 64KiB */
#define MAX_REQUEST_SIZE (64 * 1024)

int main(int argc, char **argv)
{
    int sd, err;
    struct addrinfo hints, *ptr, *res;
    rxb_t rxb;
    size_t response_len;
    char *ackVer = "ack";
    char anno[MAX_REQUEST_SIZE], N[MAX_REQUEST_SIZE], ack[10], tipologia[MAX_REQUEST_SIZE];
    /* Controllo argomenti */
    if (argc != 3)
    {
        fprintf(stderr, "Sintassi: %s server porta\n", argv[0]);
        exit(EXIT_FAILURE);
    }

    /* Preparo direttive getaddrinfo */
    memset(&hints, 0, sizeof(hints));
    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = SOCK_STREAM;

    /* Invoco getaddrinfo */
    err = getaddrinfo(argv[1], argv[2], &hints, &res);
    if (err != 0)
    {
        fprintf(stderr, "Errore risoluzione nome: %s\n", gai_strerror(err));
        exit(EXIT_FAILURE);
    }

    /* Connessione con fallback */
    for (ptr = res; ptr != NULL; ptr = ptr->ai_next)
    {
        sd = socket(ptr->ai_family, ptr->ai_socktype, ptr->ai_protocol);
        /* Se la socket fallisce, passo all'indirizzo successivo */
        if (sd < 0)
            continue;

        /* Se la connect va a buon fine, esco dal ciclo */
        if (connect(sd, ptr->ai_addr, ptr->ai_addrlen) == 0)
            break;

        /* Altrimenti, chiudo la socket e passo all'indirizzo
                 * successivo */
        close(sd);
    }

    /* Controllo che effettivamente il client sia connesso */
    if (ptr == NULL)
    {
        fprintf(stderr, "Errore di connessione!\n");
        exit(EXIT_FAILURE);
    }

    /* A questo punto, posso liberare la memoria allocata da getaddrinfo */
    freeaddrinfo(res);


    for (;;)
    {   
        //Inizio il buffer di ricezione a ogni nuvo ciclo di richieste
        rxb_init(&rxb, MAX_REQUEST_SIZE);

        puts("Inserisci l'anno di interesse: ");
        if (fgets(anno, sizeof(anno), stdin) == NULL)
        {
            perror("fgets");
            exit(EXIT_FAILURE);
        }
        if(strcmp(anno, "fine\n") == 0){
            break;
        }
        puts("Inserisci il numero di spese da visualizzare: ");
        if (fgets(N, sizeof(N), stdin) == NULL)
        {
            perror("fgets");
            exit(EXIT_FAILURE);
        }

        if(strcmp(N, "fine\n") == 0){
            break;
        }
        puts("Inserisci la tipologia dispese di interesse: ");
        if (fgets(tipologia, sizeof(tipologia), stdin) == NULL)
        {
            perror("fgets");
            exit(EXIT_FAILURE);
        }

        if(strcmp(tipologia, "fine\n") == 0){
            break;
        }
        /* Invio richieste al Server compreso il /n*/
        //Invio anno
        if (write_all(sd, anno, strlen(anno)) < 0)
        {
            perror("write");
            exit(EXIT_FAILURE);
        }
        //primo ack
        memset(ack, 0, sizeof(ack));
        response_len = sizeof(ack) - 1;
        if (rxb_readline(&rxb, sd, ack, &response_len) < 0)
        {
            rxb_destroy(&rxb);
            fprintf(stderr, "Connessione chiusa dal server!\n");
            break;
        }
        if (strcmp(ackVer, ack) != 0)
        {
            perror("Ack non corretto");
            exit(EXIT_FAILURE);
        }
        //Invio il numero di spese da visualizzare
        if (write_all(sd, N, strlen(N)) < 0)
        {
            perror("write");
            exit(EXIT_FAILURE);
        }
        //secondo ack
        memset(ack, 0, sizeof(ack));
        response_len = sizeof(ack) - 1;
        if (rxb_readline(&rxb, sd, ack, &response_len) < 0)
        {
            rxb_destroy(&rxb);
            fprintf(stderr, "Connessione chiusa dal server!\n");
            break;
        }
        //Invio la tipologia di interesse
        if (write_all(sd, tipologia, strlen(tipologia)) < 0)
        {
            perror("write");
            exit(EXIT_FAILURE);
        }
        //meglio effettuare il flush
        fflush(stdout);

        for (;;)
        {
            char response[MAX_REQUEST_SIZE];

            memset(response, 0, sizeof(response));
            response_len = sizeof(response) - 1;

            if (rxb_readline(&rxb, sd, response, &response_len) < 0)
            {
                rxb_destroy(&rxb);
                fprintf(stderr, "Connessione chiusa dal server!\n");
                exit(EXIT_FAILURE);
            }

            /* Stampo riga letta da Server */
            puts(response);

            /* Passo a nuova richiesta una volta terminato input Server */
            if (strcmp(response, "--END--") == 0)
            {
                break;
            }
        }
    }
    close(sd);
    return 0;
}